### DISCUSSION TITLE
Set embedded XMP metadata in PSD file

### DISCUSSION THREAD

**Discussion by mmarianelli-lvr** [dlemstra/Magick.NET#1517] [Help]
Hello! 
I'm trying to write embedded xmp metadata inside a psd files that doesn't contain an xmp profile.

When I set it with the SetProfile function I see that the profile is succesfully saved inside the MagickImage object but after the invocation of the Write function it seems that the xmp profile is not saved inside the file.
The same process with a jpeg file works fine.

Is there a limitation on Psd files?

Thank you

**Comment by dlemstra** [comment]
The PSD encoder/decoder of ImageMagick does not read or write the xmp profile. I even wonder if that is even supported.

**Reply by mmarianelli-lvr** [reply]
Thank you @dlemstra! I asked you because I tried to write a metadata into a psd file that already have an embedded xmp profile and it works but if the file doesn't have the embedded xmp profile it doesn't work.

Do you confirm that ImageMagick cannot 'initialize' a new xmp profile but it can write metadata in an existing xmp profile?
Thanks

**Reply by dlemstra** [reply]
Can you share the file so I have a bit more info.

**Reply by mmarianelli-lvr** [reply]
Of course!
In attachment a simple project where I simulate this case. There are multiple files in base of the case you want to test.
[PsdMetadataWriter.zip](https://github.com/dlemstra/Magick.NET/files/13898763/PsdMetadataWriter.zip)

Thanks for your support

**Reply by dlemstra** [reply]
I took a quick look at it looks like the `xmp` is embedded in the `8bim` profile. Will probably not have time to look at this until the weekend. Please let me know if you have enough information now.

**Reply by mmarianelli-lvr** [reply]
Hi @dlemstra , in the <file_without_embedded_xmp.psd> file ( that is the file where I'm not able to write the xmp profile) I don't find the xmp embedded in the 8bim profile as you say in the previous message.

However is it possible to insert/update the xmp metadata embedded in the 8pim profile by using MagickNet library?
I look forward to hearing from you.

**Reply by dlemstra** [reply]
The profile is indeed embedded in the `8bim` profile. There is no api to write 8bim values so you will need to create your own byte array where you write the `xmp` profile with the id `1060`.

**Reply by dlemstra** [reply]
I realized I could add an api for this to the `IEightBimProfile` to make this easier for you. Will need to investigate how much work this is going to take.

**Comment by mmarianelli-lvr** [comment]
> I realized I could add an api for this to the `IEightBimProfile` to make this easier for you. Will need to investigate how much work this is going to take.

Thank you @dlemstra. I tried to create my own byte array to set the id 1060 but it's a little bit complex. We are waiting your estimate about the api to do this.

**Reply by dlemstra** [reply]
It took me couple hours but I was able to add support for this. In the next release the `IEightBimProfile` interface will contain a `XmpProfile` property that you can use to set the profile.

**Reply by dlemstra** [reply]
Will update this to be methods instead to make it more clear that making changes to that profile won't update it inside the `8bim` profile.

**Reply by mmarianelli-lvr** [reply]
Thank you very much! When do you suppose the modify will be released?

**Reply by dlemstra** [reply]
I don't know when I will do a new release. Is this urgent for you?

**Comment by mmarianelli-lvr** [comment]
It's not too urgent but we would like to solve the problem to verify if it fits our needs.
Do you suggest a workaround code while we are waiting the release?

**Reply by dlemstra** [reply]
That is probably a good idea. You could copy the code (except the interfaces) from the this folder: https://github.com/dlemstra/Magick.NET/tree/main/src/Magick.NET.Core/Profiles/8Bim and put that in your own project. And then check if that works for you. The interface has been extended but the parts that add it to the image don't need the changes in the interface. You can implement the "old" one with the copy that you made of that folder.

**Comment by mmarianelli-lvr** [comment]
Hi @dlemstra , I tried to use the code you indicated me. It seems it works successfully! We wait for the new release but if the problem become urgent we use the workaround you suggest us.
Thank you!
